### PR TITLE
Return cursors at the end of extraction

### PR DESCRIPTION
**Pull request by wangch079**
#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## Related Pull Requests

* https://github.com/elastic/ent-search/pull/6441

### CODE REVIEW

**Review by RossAllenBell — DISMISSED** [review]
This makes sense to me. I think I'll actually wait for this to get merged before pursing the merge of my other folder cursoring fix.

File: lib/connectors_sdk/base/extractor.rb
Comment (wangch079) [reply lib/connectors_sdk/base/extractor.rb:33]: More and more connectors will use it when they are migrated to external connector repo.

File: lib/connectors_sdk/office365/custom_client.rb
Comment (wangch079) [inline lib/connectors_sdk/office365/custom_client.rb:137]: @RossAllenBell There will be a side effect:
E.g. if a folder has more than 100 files (without amy folder), after crawling this folder completely, `yielded` will be greater than 100, but there's no more folders in the stack. Instead of breaking the page, we will move to the next drive. This is similar to incremental sync:

https://github.com/elastic/connectors/blob/c2e41de30b804b9df4e555704bd5650963a299b7/lib/connectors_sdk/office365/custom_client.rb#L172-L175

Comment (RossAllenBell) [reply lib/connectors_sdk/office365/custom_client.rb:137]: Oh, I don't think you need that `&& stack.any?`. The scenario you describe should get handled with: https://github.com/elastic/connectors/pull/69/files#diff-f15a307a042cbd4ac669d0c8c9096c03543bab73c45d8984378a4a8bc251d56aR138.

File: lib/connectors_sdk/office365/extractor.rb
Comment (seanstory) [inline lib/connectors_sdk/office365/extractor.rb:62]: this just gave me a light-bulb moment - should we just add DEBUG logs for everywhere we're modififying and setting the `cursors`? Maybe that would help us to troubleshoot in the future, and feel good about our changes now.

File: spec/connectors_sdk/sharepoint/extractor_spec.rb
Comment (seanstory) [inline spec/connectors_sdk/sharepoint/extractor_spec.rb:296]: is it worth adding any tests to make sure that `completed` is getting set in the right scenarios? The changes here look like you made existing tests pass, but not like you're testing your changes fixed anything.

File: lib/connectors_sdk/office365/extractor.rb
Comment (RossAllenBell) [inline lib/connectors_sdk/office365/extractor.rb:61]: @wangch079 I think at least one issue with the _full_ sync is with this line. You don't want to record those cursors, which are only used by _incremental_ syncs. It's possible they get passed on to the next full sync, making it behave like an incremental sync?

**Comment by RossAllenBell** [comment]
@wangch079 has previously mentioned some issues with testing these changes that I was trying to reproduce so that we could get closer to merging this PR. I've been doing some testing with this branch and https://github.com/elastic/ent-search/pull/6441. This was the only issue I could find: https://github.com/elastic/connectors/pull/70/files#r847699965. I couldn't reproduce any issues with Incremental syncs.

With Full syncs, I noticed an issue where it started to behave like an Incremental sync, although that was only when I was accidentally running main Enterprise Search that didn't include the `completed?` changes. I think these changes might be fine once they stop persisting cursors for the Full jobs.

File: lib/connectors_sdk/base/extractor.rb
Comment (wangch079) [reply lib/connectors_sdk/base/extractor.rb:33]: @seanstory , yes this is for full/incremental syncs only.

File: lib/connectors_sdk/office365/extractor.rb
Comment (wangch079) [reply lib/connectors_sdk/office365/extractor.rb:61]: It won't. `cursors` will always be `nil` for non-incremental syncs, check [here](https://github.com/elastic/ent-search/blob/7286805237716d62a5cdf0d9c1014d32a71b2ae8/frito_pie/app/models/frito_pie/content_source.rb#L651-L658)

The reason that `cursors` are persisted for `full` sync is because we try to retrieve `cursors` from the last completed job if we can't find the last completed `incremental` job ([ref](https://github.com/elastic/ent-search/blob/7286805237716d62a5cdf0d9c1014d32a71b2ae8/frito_pie/app/models/frito_pie/content_source.rb#L652)):
```
last_job = most_recent_completed_job(Connectors::Job::INCREMENTAL) || most_recent_completed_job
```

We actually persist `cursors` for both [full/incremental](https://github.com/elastic/ent-search/blob/7286805237716d62a5cdf0d9c1014d32a71b2ae8/connectors/lib/connectors/work/index.rb#L65-L68) job and [delete](https://github.com/elastic/ent-search/blob/7286805237716d62a5cdf0d9c1014d32a71b2ae8/connectors/lib/connectors/work/purge.rb#L59-L62) job. We don't do the same for [permissions](https://github.com/elastic/ent-search/blob/7286805237716d62a5cdf0d9c1014d32a71b2ae8/connectors/lib/connectors/work/permissions.rb#L36-L38) job, I guess we just missed it.

File: lib/connectors_sdk/office365/custom_client.rb
Comment (wangch079) [reply lib/connectors_sdk/office365/custom_client.rb:137]: I guess what you are trying to do here, is to add back the current crawling folder id to the stack if it's not completed (i.e. `item_children_next_link` is present)

What I would like to prevent here is to prevent the following scenario:
1. we have crawled the drive completely, therefore there's no more folder to crawl, i.e. `stack` is empty
2. we have more than 100 items to yield

In such case, an empty stack will be saved as `page_cursor`

```
if break_after_page && yielded >= 100
  if cursors['item_children_next_link'].present?
    stack << folder_id
  end
  cursors['page_cursor'] = stack.dup
  break
end
```

File: lib/connectors_sdk/office365/extractor.rb
Comment (RossAllenBell) [reply lib/connectors_sdk/office365/extractor.rb:61]: Got it, that makes sense 👍

**Review by RossAllenBell — APPROVED** [review]

File: lib/connectors_sdk/office365/custom_client.rb
Comment (RossAllenBell) [reply lib/connectors_sdk/office365/custom_client.rb:137]: I think these changes are appropriate within this PR. The additional changes I intend to introduce should handle the scenario you describe.

**Pull request merged**
